tests/typescript_cast/__snapshots__/jsfmt.spec.js.snap

```diff
@@ -1,5 +1,210 @@
+
+  const testArrayLiteral = <PermissionsChecker<any> | undefined>["first", "second", "third"];
+
+  const testArrayLiteral2 = <PermissionsChecker<object> | undefined | number | string | boolean | null | never | object>["first", "second", "third"];
+
+  const insideFuncCall = myFunc(param1, <PermissionsChecker<any> | undefined>param2, param3)
+}
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// https://github.com/prettier/prettier/issues/4171
+function y() {
+  const fits = <Immutable.Map<string, any>>fits();
+  const fitsObjLiteral = <Immutable.Map<string, any>>{ a: "test" };
+  const fitsArrayLiteral = <Immutable.Map<string, any>>["test", "test2"];
+
+  const breakAfterCast = <Immutable.Map<string, any>>(
+    someExistingConfigMap.mergeDeep(fallbackOpts)
+  );
+
+  const stillTooLong = <
+    Immutable.Map<
+      string,
+      boolean,
+      number,
+      object,
+      null,
+      undefined,
+      any,
+      void,
+      never
+    >
+  >(
+    someExistingConfigMap.mergeDeep(fallbackOptions)
+  );
+
+  const stillTooLong2 = <
+    | Immutable.Map<
+        string,
+        boolean,
+        number,
+        object,
+        null,
+        undefined,
+        any,
+        void,
+        never
+      >
+    | undefined
+  >(
+    someExistingConfigMap.mergeDeep(fallbackOptions)
+  );
+
+  const testObjLiteral = <Immutable.Map<string, any>>{
+    property1: "myPropertyVal"
+  };
+
+  const testObjLiteral2 = <
+    Immutable.Map<
+      string,
+      any,
+      number,
+      boolean,
+      object,
+      null,
+      undefined,
+      never,
+      "extra long"
+    >
+  >{ property1: "myPropertyVal" };
+
+  const testArrayLiteral = <Immutable.Map<string, any>>[
+    "first",
+    "second",
+    "third"
+  ];
+
+  const testArrayLiteral2 = <
+    Immutable.Map<
+      string,
+      any,
+      number,
+      boolean,
+      object,
+      null,
+      undefined,
+      never,
+      "extra long"
+    >
+  >["first", "second", "third"];
+
+  const insideFuncCall = myFunc(
+    param1,
+    <Immutable.Map<string, any>>param2,
+    param3
+  );
+}
+
+// https://github.com/prettier/prettier/issues/4168
+function x() {
+  const fits = <PermissionsChecker<any> | undefined>(<any>permissions)[type];
+  const fitsObjLiteral = <PermissionsChecker<any> | undefined>{ a: "test" };
+  const fitsArrayLiteral = <PermissionsChecker<any> | undefined>["t1", "t2"];
+
+  const breakAfterCast = <PermissionsChecker<any> | undefined>(
+    (<any>permissions)[receiverType]
+  );
+
+  const stillTooLong = <
+    PermissionsChecker<object> | undefined | number | string | boolean
+  >(
+    (<any>permissions)[receiverType]
+  );
+
+  const stillTooLong2 = <
+    | PermissionsChecker<object>
+    | undefined
+    | number
+    | string
+    | boolean
+    | null
+    | never
+  >(
+    (<any>permissions)[receiverType]
+  );
```

Comment:
This looks like it would be better not to break here after the cast. Have you tried splitting this in 2 groups?

Reply:
I'm not against the suggestion that it might be better not to break in this case but I'm not sure how to implement it.

The point of the PR is essentially to make the first choice for breaking lines with casts be between the cast and the expression being cast. The hope being better output in cases where both the cast and the expression then both fit on their own lines without further breaking like in:

```ts
  const breakAfterCast = <PermissionsChecker<any> | undefined>(
    (<any>permissions)[receiverType]
  );
```

The problem in the case you're talking about is that even after breaking the cast and the expression onto their own lines, the cast still doesn't fit onto a single line and has to break more.

So what I can't figure out is a way to tell it "try to break on the outer group first, but if that doesn't make the subgroups fit onto their own line then undo the outer break and just break the subgroups." 

Any thoughts?